BUG: Added field name to error message. Closes #6232

yashmehrotra opened the pull request:
Changed error message (Added field name occurring more than once) as discussed in #6232 

![screenshot from 2015-08-25 03 02 05](https://cloud.githubusercontent.com/assets/5863972/9453106/b1915de4-4ad5-11e5-9d83-0286817a82fd.png)

yashmehrotra committed:
BUG: Added field name to error message. Closes #6232

jaimefrio commented:
We should add a test that validates the behavior.

Also, for Python 3 `PyString_AsString` is an alias for `PyBytes_AsString`, see [here](https://github.com/numpy/numpy/blob/maintenance/1.10.x/numpy/core/include/numpy/npy_3kcompat.h), which is likely not going to work very well if the input is a Python 3 Unicode string, will be surprised if you do not have to do some convert to ASCII magic in that case. But if you add the test suggested before, Travis will let us know how bad this is.

yashmehrotra commented:
@jaimefrio Ya, you were right. I have to convert Python3 Unicode string to ASCII else it fails in python3. Well, now I used [PyUnicode_AsASCIIString](https://docs.python.org/3/c-api/unicode.html#c.PyUnicode_AsASCIIString) to convert it. I first check whether its python3 or not, if it is I convert it to ASCII. I am now going to write the test and will update this PR as soon as possible.

yashmehrotra commented:
@jaimefrio I tested it with python 2.7.9 and python 3.4.3. Same for the test case written.

njsmith commented:
Do we support non-ASCII field names, and if so do they make things blow up?
(In Python 3 even variable names can legally contain non-ascii.)
.
It looks like in python 3 PyErr_Format uses PyUnicode_FromFormat which
accepts the special strings %S for "call str() on a PyObject*" and %R for
"call repr() on a PyObject*", but on py2 it uses PyString_FromFormat which
doesn't.
.
(I guess it isn't the end of the world if duplicate non-ASCII field names
trigger a misleading error message, but it would be nice.)

On Mon, Aug 24, 2015 at 5:23 PM, Yash Mehrotra <notifications@github.com>
wrote:

> @jaimefrio <https://github.com/jaimefrio> I tested it with python 2.7.9
> and python 3.4.3. Same for the test case written.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/numpy/numpy/pull/6242#issuecomment-134423140>.
>



-- 
Nathaniel J. Smith -- http://vorpus.org

yashmehrotra commented:
@njsmith I made a commit, and I used `PyUnicode_AsUTF8String`. I tested it with a few non-ascii characters and it was working fine.
Example
![screenshot from 2015-08-25 08 37 18](https://cloud.githubusercontent.com/assets/5863972/9457815/8c788242-4b04-11e5-88f6-6745c0e13cca.png)

jaimefrio commented on numpy/core/src/multiarray/descriptor.c:
```diff
@@ -509,8 +509,11 @@ _convert_from_array_descr(PyObject *obj, int align)
-            PyErr_SetString(PyExc_ValueError,
-                    "two fields with the same name");
+#if defined(NPY_PY3K)
+            name = PyUnicode_AsUTF8String(name);
```
`PyUnicode_AsUTF8String` returns a new reference, see [here](https://docs.python.org/3/c-api/unicode.html#c.PyUnicode_AsUTF8String), so you will need to `Py_DECREF(next)` afterwards. There may be other ways, but a possible solution could look like:

```
#if defined(NPY_PY3K)
            name = PyUnicode_AsUTF8String(name);
#else
            Py_INCREF(next);
#endif
            PyErr_Format(PyExc_ValueError,
                    "field '%s' occurs more than once", PyString_AsString(name));
            Py_DECREF(next);
            goto fail;
```

Also, we are assuming that `PyUnicode_AsUTF8String` will always succeed. This seems to be consistent with other uses of the same function in the NumPy code base, but I am not 100% sure it is a good practice.

jaimefrio commented:
Python 2's `dtype` barfs on unicode:

```
In [15]: np.dtype([(u"\u20b9", 'f8')])
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-15-2265ac1c74a8> in <module>()
----> 1 np.dtype([(u"\u20b9", 'f8')])

TypeError: data type not understood
```

So I guess we do not need to worry about `name` being `unicode` instead of `str`.

Python 3 is happy:

```
In [11]: np.dtype([("\u20b9", 'f8')])
Out[11]: dtype([('₹', '<f8')])
```

yashmehrotra commented:
@jaimefrio Thanks for the help.

jaimefrio commented on numpy/core/src/multiarray/descriptor.c:
```diff
@@ -509,8 +509,14 @@ _convert_from_array_descr(PyObject *obj, int align)
+#endif
+            PyErr_Format(PyExc_ValueError,
+                    "field '%s' occurs more than once", PyString_AsString(name));
+            Py_DECREF(name);
```
Come to think of it, if we only put this `Py_DECREF` in an `#if defined(NPY_PY3K)` block, the line count of the code is going to be the same, and we spare ourselves a spurious `Py_INCREF` / `Py_DECREF` pair in Python 2... Sorry for the noise.

yashmehrotra replied:
So, should I reset back to the previous commit and remove reference counting ?

yashmehrotra committed:
BUG: Added python3 compatibility, test case and handled reference counting for multiple field names error message

yashmehrotra commented:
@jaimefrio I hope the latest commit is good to go. I removed the extra `Py_INCREF`. Also, I squashed my multiple commits to 2 commits. I hope it is merge-ready now.

jaimefrio commented on numpy/core/src/multiarray/descriptor.c:
```diff
@@ -509,8 +509,12 @@ _convert_from_array_descr(PyObject *obj, int align)
-                    "two fields with the same name");
+#if defined(NPY_PY3K)
+            name = PyUnicode_AsUTF8String(name);
+            Py_DECREF(name);
```
This will not work: once you decref `next`, it could be garbage collected before you call `PyString_AsString` on it. What I had in mind on my other comment was something like:

```
#if defined(NPY_PY3K)
            name = PyUnicode_AsUTF8String(name);
#endif
            PyErr_Format(PyExc_ValueError,
                    "field '%s' occurs more than once", PyString_AsString(name));
#if defined(NPY_PY3K)
            Py_DECREF(next);
#endif
            goto fail;
```

It ain't pretty, but neither was the previous version, and at least it puts all the weirdness in the Python 3 code path, and leaves the Python 2 one nice and clean.

jaimefrio commented on numpy/core/tests/test_multiarray.py:
```diff
@@ -3554,6 +3554,13 @@ def test_field_rename(self):
+        assert_raises(ValueError, test_assign)
+
     if sys.version_info[0] >= 3:
         def test_bytes_fields(self):
```
Could you add the unicode test you showed in the comments here, for Python 3 versions only?

jaimefrio commented on numpy/core/tests/test_multiarray.py:
```diff
@@ -3554,6 +3554,13 @@ def test_field_rename(self):
         dt.names = ['p', 'q']
         assert_equal(dt.names, ['p', 'q'])
 
+    def test_multiple_field_name_occurance(self):
```
occurance --> occurrence

jaimefrio commented:
Take a look at the comments I inlined with your code, I think that should be it, and we can merge. Thanks!

yashmehrotra committed:
BUG: Added extra test for checking multiple unicode field names

yashmehrotra commented:
@jaimefrio Thanks for your feedback. I have made the changes accordingly.

jaimefrio commented:
Thanks for your patience, Yash, in it goes.

Seems to be your first contribution to NumPy: welcome, we hope to see many more! ;-)

yashmehrotra commented:
@jaimefrio Thanks Jaime. It was a good experience. Yes, its my first contibution, and yes, I plan to contribute more :smile:
